File: proto-wrapper-maven-integration-tests/src/test/java/space/alnovis/protowrapper/it/DiffIntegrationTest.java
package space.alnovis.protowrapper.it;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.EnabledIf;
import org.junit.jupiter.api.io.TempDir;
import space.alnovis.protowrapper.analyzer.ProtoAnalyzer;
import space.alnovis.protowrapper.analyzer.ProtocExecutor;
import space.alnovis.protowrapper.diff.SchemaDiff;
import space.alnovis.protowrapper.diff.model.BreakingChange;
import space.alnovis.protowrapper.diff.model.ChangeType;
import space.alnovis.protowrapper.diff.model.MessageDiff;

import java.io.File;
import java.nio.file.Path;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Integration tests for Schema Diff functionality using real proto files
 * from the integration-tests module.
 *
 * These tests use the existing v1 and v2 proto directories that contain
 * various schema differences.
 */
class DiffIntegrationTest {

    private Path protoRoot;
    private Path v1Dir;
    private Path v2Dir;

    @TempDir
    Path tempDir;

    @BeforeEach
    void setUp() {
        // Use the shared test-protos directory for generation scenarios
        protoRoot = Path.of("../test-protos/scenarios/generation");
        v1Dir = Path.of("../test-protos/scenarios/generation/v1");
        v2Dir = Path.of("../test-protos/scenarios/generation/v2");
    }

    static boolean isProtocAvailable() {
        try {
            Process process = new ProcessBuilder("protoc", "--version")
                .redirectErrorStream(true)
                .start();
            return process.waitFor() == 0;
        } catch (Exception e) {
            return false;
        }
    }

    @Test
    @EnabledIf("isProtocAvailable")
    void compareRealSchemas_detectsChanges() throws Exception {
        // Analyze both versions
        ProtocExecutor protocExecutor = new ProtocExecutor();
        ProtoAnalyzer analyzer = new ProtoAnalyzer();

        Path v1Descriptor = tempDir.resolve("v1-descriptor.pb");
        Path v2Descriptor = tempDir.resolve("v2-descriptor.pb");

        protocExecutor.generateDescriptor(v1Dir, v1Descriptor, new String[0], protoRoot);
        protocExecutor.generateDescriptor(v2Dir, v2Descriptor, new String[0], protoRoot);

        ProtoAnalyzer.VersionSchema v1Schema = analyzer.analyze(v1Descriptor, "v1");
        ProtoAnalyzer.VersionSchema v2Schema = analyzer.analyze(v2Descriptor, "v2");

        // Compare schemas
        SchemaDiff diff = SchemaDiff.compare(v1Schema, v2Schema);

        // Verify basic structure
        assertNotNull(diff);
        assertNotNull(diff.getSummary());
        assertEquals("v1", diff.getV1Name());
        assertEquals("v2", diff.getV2Name());

        // The real protos should have some differences
        // We don't assert specific counts as protos may change
        System.out.println("Diff Summary:");
        System.out.println("  Added messages: " + diff.getSummary().addedMessages());
        System.out.println("  Removed messages: " + diff.getSummary().removedMessages());
        System.out.println("  Modified messages: " + diff.getSummary().modifiedMessages());
        System.out.println("  Added enums: " + diff.getSummary().addedEnums());
        System.out.println("  Removed enums: " + diff.getSummary().removedEnums());
        System.out.println("  Modified enums: " + diff.getSummary().modifiedEnums());
    }

    @Test
    @EnabledIf("isProtocAvailable")
    void compareRealSchemas_generatesTextOutput() throws Exception {
        ProtocExecutor protocExecutor = new ProtocExecutor();
        ProtoAnalyzer analyzer = new ProtoAnalyzer();

        Path v1Descriptor = tempDir.resolve("v1-descriptor.pb");
        Path v2Descriptor = tempDir.resolve("v2-descriptor.pb");

        protocExecutor.generateDescriptor(v1Dir, v1Descriptor, new String[0], protoRoot);
        protocExecutor.generateDescriptor(v2Dir, v2Descriptor, new String[0], protoRoot);

        ProtoAnalyzer.VersionSchema v1Schema = analyzer.analyze(v1Descriptor, "v1");
        ProtoAnalyzer.VersionSchema v2Schema = analyzer.analyze(v2Descriptor, "v2");

        SchemaDiff diff = SchemaDiff.compare(v1Schema, v2Schema);

        String textOutput = diff.toText();

        assertNotNull(textOutput);
        assertTrue(textOutput.contains("Schema Comparison: v1 -> v2"));
        assertTrue(textOutput.contains("MESSAGES"));
        assertTrue(textOutput.contains("SUMMARY"));

        System.out.println("=== Text Output ===");
        System.out.println(textOutput);
    }

    @Test
    @EnabledIf("isProtocAvailable")
    void compareRealSchemas_generatesJsonOutput() throws Exception {
        ProtocExecutor protocExecutor = new ProtocExecutor();
        ProtoAnalyzer analyzer = new ProtoAnalyzer();

        Path v1Descriptor = tempDir.resolve("v1-descriptor.pb");
        Path v2Descriptor = tempDir.resolve("v2-descriptor.pb");

        protocExecutor.generateDescriptor(v1Dir, v1Descriptor, new String[0], protoRoot);
        protocExecutor.generateDescriptor(v2Dir, v2Descriptor, new String[0], protoRoot);

        ProtoAnalyzer.VersionSchema v1Schema = analyzer.analyze(v1Descriptor, "v1");
        ProtoAnalyzer.VersionSchema v2Schema = analyzer.analyze(v2Descriptor, "v2");

        SchemaDiff diff = SchemaDiff.compare(v1Schema, v2Schema);

        String jsonOutput = diff.toJson();

        assertNotNull(jsonOutput);
        assertTrue(jsonOutput.startsWith("{"));
        assertTrue(jsonOutput.contains("\"v1\": \"v1\""));
        assertTrue(jsonOutput.contains("\"v2\": \"v2\""));
        assertTrue(jsonOutput.contains("\"summary\""));
    }

    @Test
    @EnabledIf("isProtocAvailable")
    void compareRealSchemas_generatesMarkdownOutput() throws Exception {
        ProtocExecutor protocExecutor = new ProtocExecutor();
        ProtoAnalyzer analyzer = new ProtoAnalyzer();

        Path v1Descriptor = tempDir.resolve("v1-descriptor.pb");
        Path v2Descriptor = tempDir.resolve("v2-descriptor.pb");

        protocExecutor.generateDescriptor(v1Dir, v1Descriptor, new String[0], protoRoot);
        protocExecutor.generateDescriptor(v2Dir, v2Descriptor, new String[0], protoRoot);

        ProtoAnalyzer.VersionSchema v1Schema = analyzer.analyze(v1Descriptor, "v1");
        ProtoAnalyzer.VersionSchema v2Schema = analyzer.analyze(v2Descriptor, "v2");

        SchemaDiff diff = SchemaDiff.compare(v1Schema, v2Schema);

        String markdownOutput = diff.toMarkdown();

        assertNotNull(markdownOutput);
        assertTrue(markdownOutput.contains("# Schema Comparison: v1 -> v2"));
        assertTrue(markdownOutput.contains("## Summary"));
        assertTrue(markdownOutput.contains("| Category | Added | Modified | Removed |"));
    }

    @Test
    @EnabledIf("isProtocAvailable")
    void compareRealSchemas_detectsBreakingChanges() throws Exception {
        ProtocExecutor protocExecutor = new ProtocExecutor();
        ProtoAnalyzer analyzer = new ProtoAnalyzer();

        Path v1Descriptor = tempDir.resolve("v1-descriptor.pb");
        Path v2Descriptor = tempDir.resolve("v2-descriptor.pb");

        protocExecutor.generateDescriptor(v1Dir, v1Descriptor, new String[0], protoRoot);
        protocExecutor.generateDescriptor(v2Dir, v2Descriptor, new String[0], protoRoot);

        ProtoAnalyzer.VersionSchema v1Schema = analyzer.analyze(v1Descriptor, "v1");
        ProtoAnalyzer.VersionSchema v2Schema = analyzer.analyze(v2Descriptor, "v2");

        SchemaDiff diff = SchemaDiff.compare(v1Schema, v2Schema);

        // Check breaking changes API
        List<BreakingChange> errors = diff.getErrors();
        List<BreakingChange> warnings = diff.getWarnings();

        assertNotNull(errors);
        assertNotNull(warnings);

        if (diff.hasBreakingChanges()) {
            System.out.println("Breaking changes detected:");
            for (BreakingChange bc : diff.getBreakingChanges()) {
                System.out.println("  [" + bc.severity() + "] " + bc.type() + ": " +
                    bc.entityPath() + " - " + bc.description());
            }
        }
    }

    @Test
    @EnabledIf("isProtocAvailable")
    void compareRealSchemas_queryMethods() throws Exception {
        ProtocExecutor protocExecutor = new ProtocExecutor();
        ProtoAnalyzer analyzer = new ProtoAnalyzer();

        Path v1Descriptor = tempDir.resolve("v1-descriptor.pb");
        Path v2Descriptor = tempDir.resolve("v2-descriptor.pb");

        protocExecutor.generateDescriptor(v1Dir, v1Descriptor, new String[0], protoRoot);
        protocExecutor.generateDescriptor(v2Dir, v2Descriptor, new String[0], protoRoot);

        ProtoAnalyzer.VersionSchema v1Schema = analyzer.analyze(v1Descriptor, "v1");
        ProtoAnalyzer.VersionSchema v2Schema = analyzer.analyze(v2Descriptor, "v2");

        SchemaDiff diff = SchemaDiff.compare(v1Schema, v2Schema);

        // Test query methods
        assertNotNull(diff.getAddedMessages());
        assertNotNull(diff.getRemovedMessages());
        assertNotNull(diff.getModifiedMessages());
        assertNotNull(diff.getAddedEnums());
        assertNotNull(diff.getRemovedEnums());
        assertNotNull(diff.getModifiedEnums());

        // Check that modified messages have proper field changes
        for (MessageDiff md : diff.getModifiedMessages()) {
            assertNotNull(md.messageName());
            assertEquals(ChangeType.MODIFIED, md.changeType());
            assertNotNull(md.getAddedFields());
            assertNotNull(md.getRemovedFields());
            assertNotNull(md.getModifiedFields());
        }
    }
}

File: proto-wrapper-gradle-integration-tests/src/test/kotlin/io/alnovis/protowrapper/it/DiffIntegrationTest.kt
package io.alnovis.protowrapper.it

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.condition.EnabledIf
import org.junit.jupiter.api.io.TempDir
import java.io.File
import java.nio.file.Path
import java.util.concurrent.TimeUnit

/**
 * Integration tests for Schema Diff functionality.
 *
 * These tests create real Gradle projects and run actual Gradle builds.
 * The plugin must be published to mavenLocal before running these tests:
 *   ./gradlew publishToMavenLocal
 */
class DiffIntegrationTest {

    @TempDir
    lateinit var testProjectDir: Path

    private val pluginVersion: String = System.getProperty("pluginVersion", "1.5.0")
    private val projectRoot: String = System.getProperty("projectRoot", "..")
    private val testProtosDir: File = File(System.getProperty("testProtosDir", "../test-protos/scenarios"))
    private val gradleExecutable: String = System.getProperty("gradleExecutable", findGradleExecutable())

    companion object {
        @JvmStatic
        fun isProtocAvailable(): Boolean {
            return try {
                val process = ProcessBuilder("protoc", "--version")
                    .redirectErrorStream(true)
                    .start()
                process.waitFor(10, TimeUnit.SECONDS) && process.exitValue() == 0
            } catch (e: Exception) {
                false
            }
        }

        private fun findGradleExecutable(): String {
            // Check common locations for gradle
            val candidates = listOf(
                "/usr/bin/gradle",
                "/usr/local/bin/gradle",
                "${System.getProperty("user.home")}/.sdkman/candidates/gradle/current/bin/gradle",
                "gradle" // fallback to PATH
            )
            for (candidate in candidates) {
                val file = File(candidate)
                if (file.exists() && file.canExecute()) {
                    return candidate
                }
            }
            return "gradle"
        }
    }

    @BeforeEach
    fun setUp() {
        // Create settings.gradle.kts
        val settingsFile = testProjectDir.resolve("settings.gradle.kts").toFile()
        settingsFile.writeText("""
            pluginManagement {
                repositories {
                    mavenLocal()
                    gradlePluginPortal()
                    mavenCentral()
                }
            }

            dependencyResolutionManagement {
                repositories {
                    mavenLocal()
                    mavenCentral()
                }
            }

            rootProject.name = "test-project"
        """.trimIndent())
    }

    private fun copyScenarioProtos(scenarioName: String): Pair<File, File> {
        val v1Dir = testProjectDir.resolve("proto/v1").toFile()
        val v2Dir = testProjectDir.resolve("proto/v2").toFile()
        v1Dir.mkdirs()
        v2Dir.mkdirs()

        val scenarioDir = File(testProtosDir, "diff/$scenarioName")
        File(scenarioDir, "v1").listFiles()?.forEach { file ->
            file.copyTo(File(v1Dir, file.name), overwrite = true)
        }
        File(scenarioDir, "v2").listFiles()?.forEach { file ->
            file.copyTo(File(v2Dir, file.name), overwrite = true)
        }

        return v1Dir to v2Dir
    }

    private fun createBuildFile(taskConfig: String) {
        val buildFile = testProjectDir.resolve("build.gradle.kts").toFile()
        buildFile.writeText("""
            plugins {
                id("io.alnovis.proto-wrapper") version "$pluginVersion"
            }

            tasks.register<io.alnovis.protowrapper.gradle.SchemaDiffTask>("diffSchemas") {
                $taskConfig
            }
        """.trimIndent())
    }

    private fun runGradle(vararg args: String): GradleResult {
        val command = listOf(gradleExecutable, "--no-daemon", "--stacktrace") + args
        val process = ProcessBuilder(command)
            .directory(testProjectDir.toFile())
            .redirectErrorStream(true)
            .start()

        val output = process.inputStream.bufferedReader().readText()
        val completed = process.waitFor(5, TimeUnit.MINUTES)

        return if (completed) {
            GradleResult(process.exitValue(), output)
        } else {
            process.destroyForcibly()
            GradleResult(-1, "$output\n[TIMEOUT]")
        }
    }

    data class GradleResult(val exitCode: Int, val output: String) {
        val isSuccess: Boolean get() = exitCode == 0
        val isFailed: Boolean get() = exitCode != 0
    }

    // ==================== Tests ====================

    @Test
    @EnabledIf("isProtocAvailable")
    fun `diff task detects simple additions`() {
        copyScenarioProtos("01-simple-add")

        createBuildFile("""
            v1Directory.set(file("proto/v1"))
            v2Directory.set(file("proto/v2"))
            v1Name.set("v1")
            v2Name.set("v2")
            outputFormat.set("text")
            outputFile.set(file("build/diff-report.txt"))
        """.trimIndent())

        val result = runGradle("diffSchemas")

        assertThat(result.isSuccess).withFailMessage { "Build failed:\n${result.output}" }.isTrue()

        val reportFile = testProjectDir.resolve("build/diff-report.txt").toFile()
        assertThat(reportFile).exists()

        val content = reportFile.readText()
        assertThat(content).contains("Schema Comparison: v1 -> v2")
        assertThat(content).contains("ADDED")
    }

    @Test
    @EnabledIf("isProtocAvailable")
    fun `diff task detects simple removals`() {
        copyScenarioProtos("02-simple-remove")

        createBuildFile("""
            v1Directory.set(file("proto/v1"))
            v2Directory.set(file("proto/v2"))
            v1Name.set("v1")
            v2Name.set("v2")
            outputFormat.set("text")
            outputFile.set(file("build/diff-report.txt"))
        """.trimIndent())

        val result = runGradle("diffSchemas")

        assertThat(result.isSuccess).withFailMessage { "Build failed:\n${result.output}" }.isTrue()

        val reportFile = testProjectDir.resolve("build/diff-report.txt").toFile()
        assertThat(reportFile).exists()

        val content = reportFile.readText()
        assertThat(content).contains("Schema Comparison: v1 -> v2")
        assertThat(content).contains("REMOVED")
    }

    @Test
    @EnabledIf("isProtocAvailable")
    fun `diff task detects field changes`() {
        copyScenarioProtos("03-field-changes")

        createBuildFile("""
            v1Directory.set(file("proto/v1"))
            v2Directory.set(file("proto/v2"))
            v1Name.set("v1")
            v2Name.set("v2")
            outputFormat.set("text")
            outputFile.set(file("build/diff-report.txt"))
        """.trimIndent())

        val result = runGradle("diffSchemas")

        assertThat(result.isSuccess).withFailMessage { "Build failed:\n${result.output}" }.isTrue()

        val reportFile = testProjectDir.resolve("build/diff-report.txt").toFile()
        assertThat(reportFile).exists()

        val content = reportFile.readText()
        assertThat(content).contains("Schema Comparison: v1 -> v2")
        assertThat(content).contains("MODIFIED")
    }

    @Test
    @EnabledIf("isProtocAvailable")
    fun `diff task detects enum changes`() {
        copyScenarioProtos("04-enum-changes")

        createBuildFile("""
            v1Directory.set(file("proto/v1"))
            v2Directory.set(file("proto/v2"))
            v1Name.set("v1")
            v2Name.set("v2")
            outputFormat.set("text")
            outputFile.set(file("build/diff-report.txt"))
        """.trimIndent())

        val result = runGradle("diffSchemas")

        assertThat(result.isSuccess).withFailMessage { "Build failed:\n${result.output}" }.isTrue()

        val reportFile = testProjectDir.resolve("build/diff-report.txt").toFile()
        assertThat(reportFile).exists()
    }

    @Test
    @EnabledIf("isProtocAvailable")
    fun `diff task detects complex changes`() {
        copyScenarioProtos("05-complex-changes")

        createBuildFile("""
            v1Directory.set(file("proto/v1"))
            v2Directory.set(file("proto/v2"))
            v1Name.set("v1")
            v2Name.set("v2")
            outputFormat.set("text")
            outputFile.set(file("build/diff-report.txt"))
        """.trimIndent())

        val result = runGradle("diffSchemas")

        assertThat(result.isSuccess).withFailMessage { "Build failed:\n${result.output}" }.isTrue()

        val reportFile = testProjectDir.resolve("build/diff-report.txt").toFile()
        assertThat(reportFile).exists()

        val content = reportFile.readText()
        assertThat(content).contains("Schema Comparison: v1 -> v2")
    }

    @Test
    @EnabledIf("isProtocAvailable")
    fun `diff task reports no changes for identical schemas`() {
        copyScenarioProtos("06-no-changes")

        createBuildFile("""
            v1Directory.set(file("proto/v1"))
            v2Directory.set(file("proto/v2"))
            v1Name.set("v1")
            v2Name.set("v2")
            outputFormat.set("text")
            outputFile.set(file("build/diff-report.txt"))
        """.trimIndent())

        val result = runGradle("diffSchemas")

        assertThat(result.isSuccess).withFailMessage { "Build failed:\n${result.output}" }.isTrue()

        val reportFile = testProjectDir.resolve("build/diff-report.txt").toFile()
        assertThat(reportFile).exists()

        val content = reportFile.readText()
        // Check that summary shows zero changes
        assertThat(content).contains("Messages:  +0 added, ~0 modified, -0 removed")
        assertThat(content).contains("Enums:     +0 added, ~0 modified, -0 removed")
    }

    @Test
    @EnabledIf("isProtocAvailable")
    fun `diff task generates JSON output`() {
        copyScenarioProtos("03-field-changes")

        createBuildFile("""
            v1Directory.set(file("proto/v1"))
            v2Directory.set(file("proto/v2"))
            v1Name.set("v1")
            v2Name.set("v2")
            outputFormat.set("json")
            outputFile.set(file("build/diff-report.json"))
        """.trimIndent())

        val result = runGradle("diffSchemas")

        assertThat(result.isSuccess).withFailMessage { "Build failed:\n${result.output}" }.isTrue()

        val reportFile = testProjectDir.resolve("build/diff-report.json").toFile()
        assertThat(reportFile).exists()

        val content = reportFile.readText()
        assertThat(content).startsWith("{")
        assertThat(content).contains("\"v1\": \"v1\"")
        assertThat(content).contains("\"v2\": \"v2\"")
        assertThat(content).contains("\"summary\"")
    }

    @Test
    @EnabledIf("isProtocAvailable")
    fun `diff task generates Markdown output`() {
        copyScenarioProtos("03-field-changes")

        createBuildFile("""
            v1Directory.set(file("proto/v1"))
            v2Directory.set(file("proto/v2"))
            v1Name.set("v1")
            v2Name.set("v2")
            outputFormat.set("markdown")
            outputFile.set(file("build/diff-report.md"))
        """.trimIndent())

        val result = runGradle("diffSchemas")

        assertThat(result.isSuccess).withFailMessage { "Build failed:\n${result.output}" }.isTrue()

        val reportFile = testProjectDir.resolve("build/diff-report.md").toFile()
        assertThat(reportFile).exists()

        val content = reportFile.readText()
        assertThat(content).contains("# Schema Comparison: v1 -> v2")
        assertThat(content).contains("## Summary")
    }

    @Test
    @EnabledIf("isProtocAvailable")
    fun `diff task fails on breaking changes when failOnBreaking is true`() {
        copyScenarioProtos("03-field-changes")

        createBuildFile("""
            v1Directory.set(file("proto/v1"))
            v2Directory.set(file("proto/v2"))
            v1Name.set("v1")
            v2Name.set("v2")
            outputFormat.set("text")
            outputFile.set(file("build/diff-report.txt"))
            failOnBreaking.set(true)
        """.trimIndent())

        val result = runGradle("diffSchemas")

        assertThat(result.isFailed).withFailMessage { "Expected build to fail:\n${result.output}" }.isTrue()
        assertThat(result.output).contains("Breaking changes detected")
    }

    @Test
    @EnabledIf("isProtocAvailable")
    fun `diff task with breakingOnly shows only breaking changes`() {
        copyScenarioProtos("03-field-changes")

        createBuildFile("""
            v1Directory.set(file("proto/v1"))
            v2Directory.set(file("proto/v2"))
            v1Name.set("v1")
            v2Name.set("v2")
            outputFormat.set("text")
            outputFile.set(file("build/diff-report.txt"))
            breakingOnly.set(true)
        """.trimIndent())

        val result = runGradle("diffSchemas")

        assertThat(result.isSuccess).withFailMessage { "Build failed:\n${result.output}" }.isTrue()

        val reportFile = testProjectDir.resolve("build/diff-report.txt").toFile()
        assertThat(reportFile).exists()

        val content = reportFile.readText()
        assertThat(content).contains("Breaking Changes")
    }
}
